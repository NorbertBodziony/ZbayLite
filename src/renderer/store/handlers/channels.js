import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'

import {
  typeFulfilled,
  typeRejected,
  typePending,
  errorNotification,
  successNotification,
  LoaderState
} from './utils'
import notificationsHandlers from './notifications'
import identitySelectors from '../selectors/identity'
import modalsHandlers from './modals'

import { getVault } from '../../vault'
import { getClient } from '../../zcash'

export const ChannelsState = Immutable.Record({
  data: Immutable.List(),
  loader: LoaderState()
}, 'ChannelsState')

export const initialState = ChannelsState()

export const actionTypes = {
  LOAD_CHANNELS: 'LOAD_IDENTITY_CHANNELS'
}

const loadChannels = createAction(actionTypes.LOAD_CHANNELS, async (id) => {
  const channels = await getVault().channels.listChannels(id)

  await Promise.all(
    channels.map(
      channel => getClient().keys.importIVK({ ivk: channel.keys.ivk, address: channel.address })
    )
  )
  return channels
})

export const actions = {
  loadChannels
}

const _createChannel = async (identityId, { name, description }) => {
  const address = await getClient().addresses.create('sapling')
  const [ivk, sk] = await Promise.all([
    getClient().keys.exportIVK(address),
    getClient().keys.exportSK(address)
  ])
  const channel = {
    address,
    name: name,
    description: description,
    private: true,
    keys: {
      ivk,
      sk
    }
  }
  return getVault().channels.importChannel(identityId, channel)
}

const createChannel = (values, formActions) => async (dispatch, getState) => {
  const closeModal = modalsHandlers.actionCreators.closeModal('createChannel')
  try {
    const identityId = identitySelectors.id(getState())
    await _createChannel(identityId, values)
    dispatch(notificationsHandlers.actions.enqueueSnackbar(
      successNotification(`Successfully created ${values.name} channel.`)
    ))
    formActions.setSubmitting(false)
    dispatch(closeModal())
    dispatch(loadChannels(identityId))
  } catch (error) {
    dispatch(notificationsHandlers.actions.enqueueSnackbar(
      errorNotification({ message: `Failed to create channel: ${error.message}` })
    ))
    formActions.setSubmitting(false)
  }
}

export const epics = {
  createChannel
}

export const reducer = handleActions({
  [typePending(actionTypes.LOAD_CHANNELS)]: state => state.setIn(['loader', 'loading'], true)
    .setIn(['loader', 'message'], 'Loading channels'),
  [typeFulfilled(actionTypes.LOAD_CHANNELS)]: (state, { payload: data }) => state
    .set('data', Immutable.fromJS(data))
    .setIn(['loader', 'loading'], false),
  [typeRejected(actionTypes.LOAD_CHANNELS)]: (state, { payload: error }) => state
    .setIn(['loader', 'loading'], false)
}, initialState)

export default {
  actions,
  epics,
  reducer
}
