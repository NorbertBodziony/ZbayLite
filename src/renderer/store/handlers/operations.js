import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'
import * as R from 'ramda'

import { getClient } from '../../zcash'

const oneOf = (...arr) => val => R.includes(val, arr)

const isFinished = oneOf('success', 'cancelled', 'failed')

const POLLING_OFFSET = 15000

export const initialState = Immutable.Map()

export const ZcashError = Immutable.Record({
  code: null,
  message: ''
}, 'ZcashError')

export const ShieldBalanceOp = Immutable.Record({
  amount: new BigNumber(0),
  from: '',
  to: ''
})

export const PendingMessageOp = Immutable.Record({
  message: Immutable.Map(),
  channelId: ''
})

export const PendingDirectMessageOp = Immutable.Record({
  message: Immutable.Map(),
  recipientAddress: '',
  recipientUsername: '',
  offerId: ''
})

export const operationTypes = {
  shieldBalance: 'shieldBalance',
  pendingMessage: 'pendingMessage',
  pendingDirectMessage: 'pendingDirectMessage'
}

export const Operation = Immutable.Record({
  opId: '',
  type: '',
  meta: {},
  txId: null,
  error: null,
  status: 'pending'
}, 'Operation')

const addOperation = createAction('ADD_PENDING_OPERATION')
const resolveOperation = createAction('RESOLVE_PENDING_OPERATION')
const removeOperation = createAction('REMOVE_PENDING_OPERATION')

export const actions = {
  addOperation,
  resolveOperation,
  removeOperation
}

const observeOperation = ({ opId, type, meta, checkConfirmationNumber }) => async (dispatch, getState) => {
  dispatch(addOperation({ opId, type, meta }))

  const subscribe = async (callback) => {
    async function poll () {
      const {
        status,
        result: { txid: txId } = {},
        error = null
      } = await getClient().operations.getStatus(opId) || {}
      if (isFinished(status)) {
        return callback(error, { status, txId })
      } else {
        setTimeout(poll, POLLING_OFFSET)
      }
    }
    return poll()
  }

  return subscribe((error, { status, txId }) => {
    dispatch(resolveOperation({ opId, status, txId, error }))
    if (checkConfirmationNumber && !error) {
      checkConfirmationNumber({ opId, status, txId, getState, dispatch })
    }
  })
}

export const epics = {
  observeOperation
}

export const reducer = handleActions({
  [addOperation]: (state, { payload: { type, meta, opId } }) => state.set(
    opId,
    Operation({
      opId,
      type,
      meta
    })
  ),
  [resolveOperation]: (state, { payload: { opId, status, txId, error } }) => state.update(
    opId,
    m => m.merge({
      status,
      txId,
      error: error ? ZcashError(error) : null
    })
  ),
  [removeOperation]: (state, { payload: id }) => {
    if (state.has(id)) {
      return state.delete(id)
    }
    // It may be a txId
    return state.filter(pm => pm.txId !== id)
  }
}, initialState)

export default {
  actions,
  epics,
  reducer
}
