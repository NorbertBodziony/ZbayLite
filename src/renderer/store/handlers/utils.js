import Immutable from 'immutable'

import { notifierAction } from '../../components/ui/DismissSnackbarAction'

export const typePending = name => `${name}_PENDING`
export const typeFulfilled = name => `${name}_FULFILLED`
export const typeRejected = name => `${name}_REJECTED`

export const errorNotification = ({ message, options }) => ({
  message,
  options: {
    persist: false,
    variant: 'error',
    action: notifierAction,
    ...options
  }
})

export const successNotification = ({ message, options }) => ({
  message,
  options: {
    variant: 'success',
    ...options
  }
})

export const LoaderState = Immutable.Record({
  loading: false,
  message: ''
})

export default {
  typePending,
  typeFulfilled,
  typeRejected
}
