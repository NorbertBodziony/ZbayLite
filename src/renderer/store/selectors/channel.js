import { createSelector } from 'reselect'
import channelsSelector from './channels'
import identitySelectors from './identity'
import pendingMessagesSelectors from './pendingMessages'

const store = s => s

const channel = createSelector(store, state => state.get('channel'))

const data = createSelector(
  channelsSelector.data,
  channel,
  (channels, channel) => channels.find(ch => ch.get('id') === channel.id)
)

export const spentFilterValue = createSelector(channel, c => c.get('spentFilterValue', -1))

export const message = createSelector(channel, c => c.get('message'))

export const pendingMessages = createSelector(
  pendingMessagesSelectors.pendingMessages,
  channel,
  (pendingMessages, channel) => pendingMessages.filter(m => m.channelId === channel.id)
)

export const messagesMeta = createSelector(channel, c => c.messages)

const _isOwner = (address, message) => message.getIn(['sender', 'replyTo']) === address

export const messages = createSelector(
  identitySelectors.data,
  messagesMeta,
  pendingMessages,
  (identity, meta, pendingMessages) => {
    const displayableBroadcasted = meta.data.map(
      m => m.set('fromYou', _isOwner(identity.address, m))
    )
    const displayablePending = pendingMessages.map(
      m => m.get('message')
        .set('error', m.get('error'))
        .set('status', m.get('status'))
        .set('id', m.get('opId'))
        .set('fromYou', _isOwner(identity.address, m.get('message')))
    )
    return displayableBroadcasted.concat(displayablePending.values())
      .sortBy(m => m.get('createdAt'))
  })

export default {
  data,
  channel,
  spentFilterValue,
  message,
  pendingMessages,
  messages
}
