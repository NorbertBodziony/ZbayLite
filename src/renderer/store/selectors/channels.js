import { createSelector } from 'reselect'

const store = s => s

const channels = createSelector(store, state => state.get('channels'))
const data = createSelector(channels, ch => ch.data)
const loader = createSelector(channels, ch => ch.loader)
const errors = createSelector(channels, c => c.get('errors'))

export default {
  channels,
  loader,
  data,
  errors
}
