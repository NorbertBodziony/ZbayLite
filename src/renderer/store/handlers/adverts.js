import BigNumber from 'bignumber.js'

import identitySelectors from '../selectors/identity'
import channelSelectors from '../selectors/channel'
import offersHandlers from '../../store/handlers/offers'
import { messageType } from '../../zbay/messages'
import { messages } from '../../zbay'
import { getClient } from '../../zcash'
import notificationsHandlers from './notifications'
import directMessagesQueueHandlers from './directMessagesQueue'
import { errorNotification, successNotification } from './utils'
import operationsHandlers, { operationTypes } from './operations'
import contactsHandlers from './contacts'

const handleSend = ({ values }) => async (dispatch, getState) => {
  const data = {
    tag: values.tag,
    background: values.background.toString(),
    title: values.title,
    provideShipping: values.shippingInfo ? '1' : '0',
    amount: values.usd,
    description: values.description
  }
  const identityAddress = identitySelectors.address(getState())
  const channel = channelSelectors.data(getState()).toJS()
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.AD,
      data: data,
      spent: new BigNumber('0.0001')
    },
    privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: channel.address,
    identityAddress
  })
  try {
    const opId = await getClient().payment.send(transfer)
    await dispatch(
      operationsHandlers.epics.observeOperation({
        opId,
        type: operationTypes.pendingMessage,
        meta: {
          channelId: channel.id,
          message: message
        }
      })
    )
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification('Advert successfully posted')
      )
    )
  } catch (err) {
    notificationsHandlers.actions.enqueueSnackbar(
      dispatch(
        errorNotification({
          message: "Couldn't send the message, please check node connection."
        })
      )
    )
  }
}

const handleSendTransfer = ({ values, history, payload }) => async (dispatch, getState) => {
  const shippingData = identitySelectors.shippingData(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.ITEM_TRANSFER,
      data: {
        itemId: payload.id.substring(0, 64),
        tag: payload.tag,
        offerOwner: payload.offerOwner,
        shippingData
      },
      spent: new BigNumber(values.zec)
    },
    privKey
  })
  dispatch(offersHandlers.epics.createOfferAdvert({ payload, history }))
  dispatch(directMessagesQueueHandlers.epics.addDirectMessage({
    message,
    recipientAddress: payload.address,
    recipientUsername: payload.offerOwner
  })
  )
  dispatch(contactsHandlers.epics.updateDeletedChannelTimestamp({ address: payload.id + payload.offerOwner, timestamp: 0 }))
}

export const epics = {
  handleSend,
  handleSendTransfer
}

export default {
  epics
}
