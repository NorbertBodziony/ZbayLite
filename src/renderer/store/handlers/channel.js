import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'
import { createAction, handleActions } from 'redux-actions'

import operationsHandlers, { operationTypes, PendingMessageOp } from './operations'
import notificationsHandlers from './notifications'
import channelSelectors from '../selectors/channel'
import identitySelectors from '../selectors/identity'
import nodeSelectors from '../selectors/node'
import operationsSelectors from '../selectors/operations'
import { getClient } from '../../zcash'
import { messages } from '../../zbay'
import { errorNotification, LoaderState } from './utils'
import { channelToUri } from '../../zbay/channels'

export const MessagesState = Immutable.Record({
  loader: LoaderState({ loading: true }),
  data: Immutable.List(),
  errors: ''
})

export const ChannelState = Immutable.Record({
  spentFilterValue: new BigNumber(0),
  id: null,
  message: '',
  shareableUri: '',
  messages: MessagesState(),
  members: null
}, 'ChannelState')

export const initialState = ChannelState()

const setSpentFilterValue = createAction('SET_SPENT_FILTER_VALUE', (_, value) => value)
const setMessage = createAction('SET_CHANNEL_MESSAGE', R.path(['target', 'value']))
const setChannelId = createAction('SET_CHANNEL_ID')
const setMessages = createAction('SET_CHANNEL_MESSAGES')
const setLoading = createAction('SET_CHANNEL_LOADING')
const setLoadingMessage = createAction('SET_CHANNEL_LOADING_MESSAGE')
const setShareableUri = createAction('SET_CHANNEL_SHAREABLE_URI')

export const actions = {
  setLoading,
  setLoadingMessage,
  setSpentFilterValue,
  setMessage,
  setShareableUri,
  setChannelId
}

const loadMessages = () => async (dispatch, getState) => {
  const channel = channelSelectors.data(getState())
  const isTestnet = nodeSelectors.node(getState()).isTestnet
  const transfers = await getClient().payment.received(channel.get('address'))

  const pendingMessages = operationsSelectors.pendingMessages(getState())

  const msgs = await Promise.all(transfers.map(
    async (transfer) => {
      const message = await messages.transferToMessage(transfer, isTestnet)

      const pendingMessage = pendingMessages.find(
        pm => pm.txId && pm.txId === message.id)
      if (pendingMessage) {
        dispatch(
          operationsHandlers.actions.removeOperation(pendingMessage.opId)
        )
      }
      return message
    }
  ))

  dispatch(setMessages(msgs))
}

const loadChannel = (id) => async (dispatch, getState) => {
  dispatch(setLoading(true))
  dispatch(setLoadingMessage('Loading messages'))
  try {
    dispatch(setChannelId(id))

    // Calculate URI on load, that way it won't be outdated, even if someone decides
    // to update channel in vault manually
    const channel = channelSelectors.data(getState())
    const uri = await channelToUri(channel.toJS())
    dispatch(setShareableUri(uri))

    await dispatch(loadMessages())
  } catch (err) {}
  dispatch(setLoading(false))
}

const sendOnEnter = (event) => async (dispatch, getState) => {
  const enterPressed = event.nativeEvent.keyCode === 13
  const shiftPressed = event.nativeEvent.shiftKey === true
  if (enterPressed && !shiftPressed) {
    event.preventDefault()
    const message = messages.createMessage({
      identity: identitySelectors.data(getState()).toJS(),
      messageData: {
        type: messages.messageType.BASIC,
        data: event.target.value
      }
    })
    const channel = channelSelectors.data(getState()).toJS()
    const transfer = await messages.messageToTransfer({
      message,
      channel
    })
    try {
      const opId = await getClient().payment.send(transfer)
      await dispatch(operationsHandlers.epics.observeOperation({
        opId,
        type: operationTypes.pendingMessage,
        meta: PendingMessageOp({
          channelId: channel.id,
          message: Immutable.fromJS(message)
        })
      }))
    } catch (err) {
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: 'Couldn\'t send the message, please check node connection.'
        })
      )
    }
    dispatch(setMessage(''))
  }
}

const resendMessage = (message) => async (dispatch, getState) => {
  dispatch(operationsHandlers.actions.removeOperation(message.id))
  const channel = channelSelectors.data(getState()).toJS()
  const transfer = await messages.messageToTransfer({
    message,
    channel
  })
  try {
    const opId = await getClient().payment.send(transfer)
    await dispatch(operationsHandlers.epics.observeOperation({
      opId,
      type: operationTypes.pendingMessage,
      meta: PendingMessageOp({
        channelId: channel.id,
        message: Immutable.fromJS(message)
      })
    }))
  } catch (err) {
    notificationsHandlers.actions.enqueueSnackbar(
      errorNotification({
        message: 'Couldn\'t send the message, please check node connection.'
      })
    )
  }
}

export const epics = {
  sendOnEnter,
  loadChannel,
  resendMessage,
  loadMessages
}

// TODO: we should have a global loader map
export const reducer = handleActions({
  [setLoading]: (state, { payload: loading }) => state.setIn(['messages', 'loader', 'loading'], loading),
  [setLoadingMessage]: (state, { payload: message }) => state.setIn(['messages', 'loader', 'message'], message),
  [setSpentFilterValue]: (state, { payload: value }) => state.set('spentFilterValue', new BigNumber(value)),
  [setMessage]: (state, { payload: value }) => state.set('message', value),
  [setChannelId]: (state, { payload: id }) => state.set('id', id),
  [setShareableUri]: (state, { payload: uri }) => state.set('shareableUri', uri),
  [setMessages]: (state, { payload: messages }) => state.update(
    'messages',
    msgMeta => msgMeta
      .set('data', Immutable.fromJS(messages))
  )
}, initialState)

export default {
  reducer,
  epics,
  actions
}
