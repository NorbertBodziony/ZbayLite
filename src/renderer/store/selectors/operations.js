import { createSelector } from 'reselect'
import { operationTypes } from '../handlers/operations'

const store = s => s

const operations = createSelector(store, state => state.get('operations'))

const pendingMessages = createSelector(
  operations,
  ops => ops.filter(o => o.type === operationTypes.pendingMessage)
)

const pendingDirectMessages = createSelector(
  operations,
  ops => ops.filter(o => o.type === operationTypes.pendingDirectMessage)
)

export default {
  operations,
  pendingMessages,
  pendingDirectMessages
}
