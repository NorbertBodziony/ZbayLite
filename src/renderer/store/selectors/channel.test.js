/* eslint import/first: 0 */
jest.mock('../../vault')
import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import * as R from 'ramda'

import channelSelectors from './channel'
import { operationTypes, PendingMessageOp, Operation } from '../handlers/operations'
import create from '../create'
import { ChannelState, MessagesState } from '../handlers/channel'
import { ChannelsState } from '../handlers/channels'
import { createMessage, createChannel, now } from '../../testUtils'
import { LoaderState } from '../handlers/utils'

describe('Channel selector', () => {
  const channelId = 'this-is-a-test-id'

  let store = null
  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      initialState: Immutable.Map({
        channel: ChannelState({
          spentFilterValue: 38,
          id: channelId,
          members: new BigNumber(0),
          message: 'Message written in the input',
          messages: MessagesState({
            data: Immutable.fromJS(
              R.range(0, 4)
                .map(i => createMessage(i, now.minus({ hours: 2 * i }).toSeconds()))
            ),
            loader: LoaderState({
              message: 'Test loading message',
              loading: true
            })
          })
        }),
        channels: ChannelsState({
          data: Immutable.fromJS([createChannel(channelId)])
        }),
        operations: Immutable.fromJS({
          'test-operation-id': Operation({
            opId: 'test-operation-id',
            txId: 'transaction-id',
            type: operationTypes.pendingMessage,
            meta: PendingMessageOp({
              message: Immutable.fromJS(createMessage(
                'test-message-id',
                now.minus({ hours: 1 }).toSeconds()
              )),
              channelId
            }),
            status: 'success'
          }),
          'test-operation-id-2': Operation({
            opId: 'test-operation-id-2',
            txId: 'transaction-id-2',
            type: operationTypes.pendingMessage,
            meta: PendingMessageOp({
              message: Immutable.fromJS(createMessage(
                'test-message-id-2',
                now.minus({ hours: 3 }).toSeconds()
              )),
              channelId: `not-${channelId}`
            }),
            status: 'success'
          }),
          'test-operation-id-3': Operation({
            opId: 'test-operation-id-3',
            txId: 'transaction-id-3',
            type: operationTypes.pendingMessage,
            meta: PendingMessageOp({
              message: Immutable.fromJS(createMessage(
                'test-message-id-3',
                now.minus({ hours: 5 }).toSeconds()
              )),
              channelId
            }),
            status: 'success'
          })
        })
      })
    })
  })

  it('channel selector', async () => {
    expect(channelSelectors.channel(store.getState())).toMatchSnapshot()
  })

  it('spent filter value selector', async () => {
    expect(channelSelectors.spentFilterValue(store.getState())).toMatchSnapshot()
  })

  it('message selector', async () => {
    expect(channelSelectors.message(store.getState())).toMatchSnapshot()
  })

  it('messages selector', async () => {
    expect(channelSelectors.messages(store.getState())).toMatchSnapshot()
  })

  it('data selector', async () => {
    expect(channelSelectors.data(store.getState())).toMatchSnapshot()
  })

  it('pendingMessages', () => {
    expect(channelSelectors.pendingMessages(store.getState())).toMatchSnapshot()
  })

  it('loader', () => {
    expect(channelSelectors.loader(store.getState())).toMatchSnapshot()
  })
})
