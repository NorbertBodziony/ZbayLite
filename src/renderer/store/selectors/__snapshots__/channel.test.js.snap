// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel selector channel selector 1`] = `
Immutable.ChannelState {
  "spentFilterValue": 38,
  "id": "this-is-a-test-id",
  "message": "Message written in the input",
  "messages": Immutable.Record {
    "loader": Immutable.Record {
      "loading": true,
      "message": "Test loading message",
    },
    "data": Immutable.List [
      Immutable.Map {
        "id": 0,
        "type": 1,
        "createdAt": 1551963828,
        "message": "This is some message 0",
        "spent": "23.23",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      Immutable.Map {
        "id": 1,
        "type": 1,
        "createdAt": 1551956628,
        "message": "This is some message 1",
        "spent": "23.23",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      Immutable.Map {
        "id": 2,
        "type": 1,
        "createdAt": 1551949428,
        "message": "This is some message 2",
        "spent": "23.23",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      Immutable.Map {
        "id": 3,
        "type": 1,
        "createdAt": 1551942228,
        "message": "This is some message 3",
        "spent": "23.23",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
    ],
    "errors": "",
  },
  "members": "0",
}
`;

exports[`Channel selector data selector 1`] = `
Immutable.Map {
  "id": "this-is-a-test-id",
  "name": "Channel this-is-a-test-id",
  "private": false,
  "hash": "test-hash-this-is-a-test-id",
  "address": "zs1testaddressthis-is-a-test-id",
  "unread": "this-is-a-test-id",
  "description": "Channel about this-is-a-test-id",
  "keys": Immutable.Map {
    "ivk": "incoming-viewing-key-this-is-a-test-id",
  },
}
`;

exports[`Channel selector loader 1`] = `
Immutable.Record {
  "loading": true,
  "message": "Test loading message",
}
`;

exports[`Channel selector message selector 1`] = `"Message written in the input"`;

exports[`Channel selector messages selector 1`] = `
Immutable.List [
  Immutable.Map {
    "id": 3,
    "type": 1,
    "createdAt": 1551942228,
    "message": "This is some message 3",
    "spent": "23.23",
    "sender": Immutable.Map {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    },
    "fromYou": false,
  },
  Immutable.Map {
    "message": "This is some message test-message-id-3",
    "error": null,
    "spent": "23.23",
    "status": "success",
    "fromYou": false,
    "sender": Immutable.Map {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    },
    "type": 1,
    "id": "test-operation-id-3",
    "createdAt": 1551945828,
  },
  Immutable.Map {
    "id": 2,
    "type": 1,
    "createdAt": 1551949428,
    "message": "This is some message 2",
    "spent": "23.23",
    "sender": Immutable.Map {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    },
    "fromYou": false,
  },
  Immutable.Map {
    "id": 1,
    "type": 1,
    "createdAt": 1551956628,
    "message": "This is some message 1",
    "spent": "23.23",
    "sender": Immutable.Map {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    },
    "fromYou": false,
  },
  Immutable.Map {
    "message": "This is some message test-message-id",
    "error": null,
    "spent": "23.23",
    "status": "success",
    "fromYou": false,
    "sender": Immutable.Map {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    },
    "type": 1,
    "id": "test-operation-id",
    "createdAt": 1551960228,
  },
  Immutable.Map {
    "id": 0,
    "type": 1,
    "createdAt": 1551963828,
    "message": "This is some message 0",
    "spent": "23.23",
    "sender": Immutable.Map {
      "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
    },
    "fromYou": false,
  },
]
`;

exports[`Channel selector pendingMessages 1`] = `
Immutable.Map {
  "test-operation-id": Immutable.Operation {
    "opId": "test-operation-id",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id",
        "type": 1,
        "createdAt": 1551960228,
        "message": "This is some message test-message-id",
        "spent": "23.23",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": "this-is-a-test-id",
    },
    "txId": "transaction-id",
    "error": null,
    "status": "success",
  },
  "test-operation-id-3": Immutable.Operation {
    "opId": "test-operation-id-3",
    "type": "pendingMessage",
    "meta": Immutable.Record {
      "message": Immutable.Map {
        "id": "test-message-id-3",
        "type": 1,
        "createdAt": 1551945828,
        "message": "This is some message test-message-id-3",
        "spent": "23.23",
        "sender": Immutable.Map {
          "replyTo": "zs1z7rejlpsa98s2rrrfkwmaxu53e4ue0ulcrw0h4x5g8jl04tak0d3mm47vdtahatqrlkngh9slya",
        },
      },
      "channelId": "this-is-a-test-id",
    },
    "txId": "transaction-id-3",
    "error": null,
    "status": "success",
  },
}
`;

exports[`Channel selector spent filter value selector 1`] = `38`;
